### ISSUE TITLE
Lexer counts lines incorrectly in CRLF files on LF/NL-only systems

### ISSUE THREAD

**Issue by jordanbrown0** [openscad/openscad#4882]
**Describe the bug**
When a file has lines ending with CRLF (\r\n, 0D 0A) on a newline (LF)-only system like Linux, the lexer counts a line each for the CR and the LF.  Except that CR in a comment isn't counted, so line numbers aren't simply doubled.

**To Reproduce**
* Create a `.scad` file with lines that end with CRLF, that generates warnings or errors.
* Run it on a Linux system.
* Observe that the line numbers reported do not match the source.

**Expected behavior**
Errors and warnings should report the correct line.

**Comment by jordanbrown0** [comment]
```
[\n\r]                                  { LOCATION_ADD_LINES(parserlloc, yyleng); }
```
tallies a line for either a LF (\n) or a CR (\r), and so tallies two lines for CRLF.

Similar patterns in comment contexts match only LF (\n).

Doesn't misbehave on Windows (where CRLF is the norm) because there the library strips CR out of text files while reading.

**Comment by pca006132** [comment]
I think CRLF is \r\n, I never saw \n\r counted as 1 line break before... Maybe I don't deal with Windows often enough.

**Comment by nophead** [comment]
Shouldn't this sort of issue be sorted out with a .gitattributes file?

For example for NopSCADlib I have this:
```
# Set the default behavior, in case people don't have core.autocrlf set.
* text=auto eol=lf

# Declare text files that are normalized and converted on checkout.
*.scad text
*.py   text
*.md   text

# Handle Windows batch and command files
*.{bat,[bB][aA][tT]} text eol=crlf
*.{cmd,[cC][mM][dD]} text eol=crlf

# Denote files that are binary and should not be modified.
*.png binary
*.jpg binary
```

**Comment by jordanbrown0** [comment]
> I think CRLF is \r\n,

Yes, it is.  But the line I quoted above (`[\n\r]`)doesn't match CRLF per se.  It's a regular expression and matches **either** `\n` (LF/NL) **or** `\r` (CR).  Thus CRLF matches that line twice.

**Comment by jordanbrown0** [comment]
I don't know whether the lexer needs to pay attention to CRs, or if it can just ignore them.  A few historical systems (notably Classic Mac OS, pre-Mac OS X) had only CR between lines, the inverse of the UNIX answer.  I don't know whether any such systems still exist, or whether we care about them.

If we can ignore CR-only formats, the simple answer is to ignore CR entirely.

If we want to allow CR-only formats, it's trickier because you want CR to count as a line break, and you want LF to count as a line break, but you want CRLF to count as one line break, not two.  I suspect that the lexer can do that relatively easily, because REs are happy to greedily suck up additional characters to match a longer pattern.

**Comment by jordanbrown0** [comment]
https://en.wikipedia.org/wiki/Newline lists these systems as CR-only:
* Commodore 8-bit machines (C64, C128)
* Acorn BBC
* ZX Spectrum
* TRS-80
* Apple II series
* Oberon
* the classic Mac OS
* MIT Lisp Machine
* OS-9

**Comment by nophead** [comment]
I think you just open text files in text mode and then the OS converts it to \n only in memory but reads and writes files in the native mode, so \r\n on Windows and \n on Linux.

And git works the same way if you set up .gitattributes correctly. I.e. local copies of text files have the line endings appropriate to your operating system.

**Comment by jordanbrown0** [comment]
@kintel I don't think there's any real disagreement on the question.  Simply nothing on Windows would display `X\r\nY\r\nZ` on anything other than three lines.  (A few things might separate those lines more than normal, treating the line breaks as paragraph breaks.)  WordPad isn't exactly a text editor; it's a small word processor.  Its native formats are not plain-text formats.  It can save in plain text, but that strips formatting.  And when it does save in plain text, it saves with \r\n.  Yes, the line numbers absolutely must match what's shown in the OpenSCAD editor.  And what happens is that if you are on Linux and have a Windows-style CRLF file, the editor shows it "correctly", and the line numbers in the editor match the line numbers in error messages, but if you run it from the command line the line numbers do *not* match.  The line numbers reported for a file loaded into the editor do not match the line numbers reported for the same file when `include<>`d or `use<>`d.  For super extra fun, they aren't even consistent - parser errors report correctly, but runtime errors report incorrectly.

@nophead Yes... but by default UNIX programs don't do anything with CR.  It's just another character, not associated with line breaks.  The libraries don't do anything with it.  OpenSCAD always mostly ignores it, but sometimes counts it as starting a new line and sometimes doesn't.

Test case 1:
```
echo("hello");
echo(nonexist);
```
On Linux, save it as a CRLF file. (vim's `set ff=dos` is helpful here.) Note the undefined variable on what I think we would all agree is line 2.

In the GUI:
```
WARNING: Ignoring unknown variable 'nonexist' in file tc1.scad, line 2
```
In the CLI:
```
WARNING: Ignoring unknown variable 'nonexist' in file tc1.scad, line 3
```

Let's add tc2.scad.  (Doesn't matter whether it is in CRLF format.)
```
include <tc1.scad>
```

GUI says:
```
WARNING: Ignoring unknown variable 'nonexist' in file tc1.scad, line 3
```

But let's change it to a syntax error.  Here's tc3.scad, in CRLF form:

```
echo("line 1");
echo("line 2");
echo("line 3");
module;
echo("line 5");
```
CLI and GUI both correctly report
```
ERROR: Parser error: syntax error in file tc3.scad, line 4
```

I have not identified why parser errors are correct and runtime errors are wrong.

**Comment by gsohler** [comment]
Whatever Line Numbers scheme IS correct, IT should correspond to what you
can Count in the source Code Window ...

Jordan Brown ***@***.***> schrieb am Fr., 8. Dez. 2023, 18:18:

**Comment by jordanbrown0** [comment]
>Whatever Line Numbers scheme IS correct, IT should correspond to what you
can Count in the source Code Window ...

Absolutely.  And, for files with CRLF on Linux, that's only true for files directly executed from the GUI.  Files executed from the CLI and from `use`/`include` in the GUI have runtime errors shown wrong.

**Comment by pca006132** [comment]
IMO we could just look at the lexer of other languages and see how they handle it.

**Comment by jordanbrown0** [comment]
This really, deeply, is not a complicated problem.  It's not even all that complicated if you want to support CR-only platforms, which I don't think we need to do.

I'll come up with a PR.

**Comment by pca006132** [comment]
I thought we can just use negative lookahead for \r in the regex, just no idea why this thread is so long...

**Comment by nophead** [comment]
You shouldn't have any windows CRLF files on Linux if you have git set up correctly. But if you just remove all the CRs then it should work if you do.

**Comment by jordanbrown0** [comment]
If I e-mail you one of my .SCAD files, it will have CRLFs.  If I e-mail you a ZIP of 20 or 30 files, do you want to have to filter the CRs out of all of them?  (And then it will be difficult to check whether you have the same files that I do.)

The lexer tries to handle both NL files and CRLF files.  It just gets it wrong.  This isn't a hard bug.

**Comment by jordanbrown0** [comment]
This is still on my list to fix when I find enough round tuits.
